posts/saving-and-analyzing-trending-topics-on-twitter-using-aws-athena-lambda-and-cdk.md

```diff
@@ -0,0 +1,198 @@
+
+# Deployment using AWS CDK
+
+[AWS Cloud Development Kit](https://aws.amazon.com/cdk/) is THE framework that you need to use to define your AWS cloud infrastructure in code. You can pick your favorite language out of the 5 supported languages (JavaScript, TypeScript, Python, Java or .NET) and its simplicity will save you a lot of time. I recommend using TypeScript since the CDK team prioritizes it and you will also have more examples available on this [GitHub repository](https://github.com/aws-samples/aws-cdk-examples).
+
+For this blog, I decided to use Python since I'm more familiar with it and tend to prefer it over TypeScript. I recommend reading this [CDK tutorial](https://docs.aws.amazon.com/cdk/latest/guide/getting_started.html) in order to have all the necessary tools installed.
+
+## Lambda
+
+In order to use the Twitter API, you will first have to create an application in order to acquire secrets/tokens. You can follow the [instructions here](https://python-twitter.readthedocs.io/en/latest/getting_started.html) on how to create a Twitter application that will use the API.
+
+The lambda code below is pretty straightforward since its task is to call the Twitter API and save the result to S3.
+
+```python
+import datetime
+import json
+import os
+
+import boto3
+import twitter
+
+s3 = boto3.resource('s3')
+
+
+def main(event, context):
+    # Create Twitter API instance
+    api = twitter.Api(consumer_key=os.environ['CONSUMER_KEY'],
+                      consumer_secret=os.environ['CONSUMER_SECRET'],
+                      access_token_key=os.environ['ACCESS_TOKEN_KEY'],
+                      access_token_secret=os.environ['ACCESS_TOKEN_SECRET'])
+
+    # Get top 10 trending topics for a specific location
+    trends = api.GetTrendsWoeid(os.environ['WOEID'])
+
+    # Format to have a JSON document on each line
+    body = '\n'.join(list(map(lambda trend: json.dumps(trend._json), trends)))
+
+    # Save to S3 with a key like "2020-07-23 05:25:08"
+    key = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
+    s3.Bucket(os.environ['BUCKET']).put_object(Key=key, Body=body)
+
+    return {
+        'message': 'UPLOAD OK'
+    }
+```
+
+## Stack
+
+The benefits of CDK is that you can define your whole infrastructure with few lines of code and have all the permissions configured. Below you have the code that create both the S3 bucket, the lambda and set a rule to execute the lambda every 15mins. The lambda's environment variables are used to pass the credentials, the S3 bucket and the WOEID (the location used to get the trending topics from). New York is used for this blog but you an replace the WOEID with one of your choice, [here is the list](https://nations24.com/world-wide) of what is available.
```

Comment:
The benefit of CDK is that you can define your whole infrastructure with merely a few lines of code and have all the permissions configured.